airflow/models/taskinstance.py

```diff
@@ -1270,23 +1333,24 @@ def check_and_change_state_before_execution(
                 ignore_task_deps=ignore_task_deps,
                 description="non-requeueable deps",
             )
-            if not self.are_dependencies_met(
+            if not ti.are_dependencies_met(
                 dep_context=non_requeueable_dep_context, session=session, verbose=True
             ):
                 session.commit()
-                return False
+                make_transient(ti)
```

Comment:
Task instance is expired if the session was not passed down from the parent caller but provided by `@provide_session`. We need to `make_transient` to be able to do:
```
ti = TI.check_and_change_state_before_execution()
```

We have only 1 case where there is no session in the parent, `LocalTaskJob._execute`.